python_modules/dagster/dagster_tests/asset_defs_tests/asset_graph_view_tests/test_bfs.py

```diff
@@ -0,0 +1,253 @@
+from dagster import AssetKey, AssetSpec, Definitions, asset, multi_asset
+from dagster._core.asset_graph_view.asset_graph_view import AssetGraphView
+from dagster._core.asset_graph_view.bfs import (
+    AssetGraphViewBfsFilterConditionResult,
+    bfs_filter_asset_graph_view,
+)
+from dagster._core.definitions.asset_graph_subset import AssetGraphSubset
+from dagster._core.definitions.events import AssetKeyPartitionKey
+from dagster._core.definitions.partition_mapping import IdentityPartitionMapping
+
+
+def _get_subset_with_keys(graph_view, keys):
+    return AssetGraphSubset.from_entity_subsets(
+        [graph_view.get_full_subset(key=key) for key in keys]
+    )
+
+
+def test_bfs_filter_empty_graph():
+    """Test BFS filter on empty graph returns empty result."""
+    graph_view = AssetGraphView.for_test(Definitions())
+    initial_subset = AssetGraphSubset.empty()
+
+    def condition_fn(subset, _visited):
+        return AssetGraphViewBfsFilterConditionResult(subset, [])
+
+    result, failed = bfs_filter_asset_graph_view(
+        graph_view, condition_fn, initial_subset, include_full_execution_set=False
+    )
+
+    assert result == AssetGraphSubset.empty()
+    assert failed == []
+
+
+def test_bfs_filter_dependency_chain():
+    """Test BFS filter on linear dependency chain."""
+
+    @asset
+    def asset1():
+        return 1
+
+    @asset
+    def asset2(asset1):
+        return asset1 + 1
+
+    @asset
+    def asset3(asset2):
+        return asset2 + 1
+
+    @asset
+    def asset4(asset3):
+        return asset3 + 1
+
+    graph_view = AssetGraphView.for_test(Definitions(assets=[asset1, asset2, asset3, asset4]))
+    initial_subset = AssetGraphSubset.from_entity_subsets(
+        [graph_view.get_full_subset(key=asset1.key)]
+    )
+
+    def condition_fn(subset, visited):
+        # Only allow asset1 and asset3
+        if AssetKey("asset3") in subset.asset_keys:
+            assert visited == AssetGraphSubset.from_entity_subsets(
+                [
+                    graph_view.get_full_subset(key=asset1.key),
+                    graph_view.get_full_subset(key=asset2.key),
+                ]
+            )
+
+            return AssetGraphViewBfsFilterConditionResult(
+                AssetGraphSubset.empty(),
+                [(subset, "asset3 not allowed")],
```

Comment:
This contradicts the comment saying that only asset1 and asset3 are allowed. Should it be asset1 and asset2 are allowed?

Reply:
updated comment